File: 3-longest-substring-without-repeating-characters/main.java
class Solution {
  public int lengthOfLongestSubstring(String s) {
    boolean[] lut = new boolean[256]; // java learn
    int left = 0, right = 0;
    int longest = 0;
    while (right != s.length()) { // java learn
      while (lut[s.charAt(right)]) { //java learn
        lut[s.charAt(left)] = false;
        left++;
      }
      lut[s.charAt(right)] = true;
      
      longest = Math.max(longest, right - left + 1); //java learn
      
      right++;
    }
    return longest;
  }
}

File: 3-longest-substring-without-repeating-characters/main.cpp
class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    if (s.empty()) return 0;
    if (s.length() == 1) return 1;
    array<bool, CHAR_MAX> lut {{false}};
    
    int left = 0, right = 0;
    int longest = 1;
    int length = 0;
    while (right != s.length()) {
      while (lut[s[right]]) {
        lut[s[left]] = false;
        length--;
        left++;
      }
      lut[s[right]] = true;
      length++;
      
      if (longest < length) longest = length;
      right ++;
    }
    return longest;
  }
};